daemon/images/image_changes.go

```diff
@@ -15,5 +17,9 @@ func (i *ImageService) Changes(ctx context.Context, container *container.Contain
 	if container.RWLayer == nil {
 		return nil, errors.New("RWLayer of container " + container.Name + " is unexpectedly nil")
 	}
-	return container.RWLayer.Changes()
+	rwLayer, ok := container.RWLayer.(layer.RWLayer)
+	if !ok {
+		return nil, fmt.Errorf("container %s has an unexpected RWLayer type: %T", container.Name, container.RWLayer)
+	}
+	return rwLayer.Changes()
```

Comment:
Ideally, we'd match this at compile-time, but I guess that's complicated because importing `layer` into the `container` package was exactly the part you were trying to avoid 🤔

Reply:
Hmm, I think we could remove the `Changes` from the image service completely and instead keep the `Changes` method in the `container.RWLayer` interface.

Reply:
I think, taking a few steps back, our goal effectively is;

- We have some methods accepting a `layer.RWLayer` interface
- But that's a rather large interface, including bits that may not be needed by those methods accepting them
- So we want to reduce the interface they require, so that we can split things up, and in some cases, reduce what has to be implemented

Reply:
Yes, right.